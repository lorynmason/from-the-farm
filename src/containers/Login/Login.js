import React, { Component } from 'react';
import { connect } from 'react-redux';
import { loginUser } from '../../thunks/loginUser';
import { postUser } from '../../thunks/postUser';
import { Redirect } from 'react-router';
import PropTypes from 'prop-types';
import ReactFilestack from 'react-filestack';
import { apikey } from '../../apikey';

export class Login extends Component {
  constructor(){
    super()
    this.state = {
      name: '',
      email: '',
      password: '',
      password_confirmation: '',
      account_type: "vendor",  
      address: '',
      city: '',
      state: '',
      phone: '',
      zip: '',
      bio: '',
      img_url: '',
      newUser: false
    }
  }

  handleChange = (e) => {
    this.setState({
      [e.target.name]: e.target.value
    })
  }

  handleLogin = (e) => {
    e.preventDefault()
    const { email, password } = this.state
    this.props.loginUser('https://xpoll-be.herokuapp.com/api/v1/authentication', { email, 
    password })
  }

  handleNewUser = () => {
    this.setState({
      newUser: !this.state.newUser
    })
  }

  handleSignup = (e) => {
    e.preventDefault()
    this.props.postUser('https://xpoll-be.herokuapp.com/api/v1/registration', this.state)
  }

  basicOptions = {
    accept: 'image/*',
    fromSources: ['local_file_system'],
    maxSize: 1024 * 1024,
    maxFiles: 1,
  }

  onSuccess = (result) => {
    console.error(result.filesUploaded[0].url); 
    this.setState({
      img_url: result.filesUploaded[0].url
    })   
  }

  onError = (error) => {
    console.error('error', error);
  }

  render() {
    let page;
    if (this.props.user.name) {
      page = <Redirect to='/profile' />
    }
    if (this.state.newUser) {
      return (
        <form className="login signup" onChange={this.handleChange} onSubmit={this.handleSignup}>
          <h3>Sign Up</h3>
          <input placeholder="Company Name" name="name" type="text" onChange={this.handleChange} value={this.state.name}/>
          <input placeholder="Email" name="email" type="email" onChange={this.handleChange} value={this.state.email}/>
          <input placeholder="Address" name="address" type="text" onChange={this.handleChange} value={this.state.address}/>
          <input placeholder="City" name="city" type="text" onChange={this.handleChange} value={this.state.city}/>
          <input placeholder="State" name="state" type="text" onChange={this.handleChange} value={this.state.state}/>
          <input placeholder="Zip Code" name="zip" type="text" onChange={this.handleChange} value={this.state.zip}/>
          <input placeholder="Phone Number" name="phone" type="text" onChange={this.handleChange} value={this.state.phone}/>
          <input placeholder="Company Bio" name="bio" type="text" onChange={this.handleChange} value={this.state.bio}/>
          <p>Add a profile picture (optional)</p>
          <ReactFilestack
            apikey={apikey}
            buttonText="Upload Photo"
            buttonClass="upload"
            options={this.basicOptions}
            onSuccess={this.onSuccess}
            onError={this.onError}
          />
          <input placeholder="Password" name="password" type="password" onChange={this.handleChange} value={this.state.password}/>
          <input placeholder="Confirm Password" name="password_confirmation" type="password" onChange={this.handleChange} value={this.state.confirmPassword}/>        
          <button>Sign Up</button>
          <p className="create-account" onClick={this.handleNewUser}>
            Login as existing user
          </p>
          {page}
        </form>
      )
    }

    return (
      <form className="login" onChange={this.handleChange} onSubmit={this.handleLogin}>
        <h3>Login</h3>
        <input placeholder="email" name="email" type="email" onChange={this.handleChange} value={this.state.email}/>
        <input placeholder="password" name="password" type="password" onChange={this.handleChange} value={this.state.password}/>        
        <button>Login</button>
        <p className="create-account" onClick={this.handleNewUser}>
          Click here to create an account
        </p>
        {page}
      </form>
    )
  }
}

export const mapStateToProps = (state) => ({
  user: state.user
});

export const mapDispatchToProps = (dispatch) => ({
  loginUser: (url, state) => dispatch(loginUser(url, state)),
  postUser: (url, state) => dispatch(postUser(url, state))
});

export default connect(mapStateToProps, mapDispatchToProps)(Login);

Login.propTypes = {
  user: PropTypes.object,
  loginUser: PropTypes.func
}
