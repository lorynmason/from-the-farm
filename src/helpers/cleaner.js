const formatPhoneNumber = (num) => {
  const phoneRegex = /^\(?([0-9]{3})\)?[-. ]?([0-9]{3})[-. ]?([0-9]{4})$/;
  return num.toString().replace(phoneRegex, "($1) $2-$3")
}

export const cleanProducts = (vendors) => {
  return vendors.reduce((productArr, vendor) => {
    const { name, products } = vendor.attributes;
    products.reduce((obj, product) => {
      const itemName = Object.keys(product)[0];

      obj = {
        vendorName: name,
        name: Object.keys(product)[0],
        ...product[itemName]
      }
      productArr.push(obj);
      return obj
    }, {})

    return productArr;
  }, [])
}

export const cleanVendors = (vendors) => {
  return vendors.map( vendor => {
    const { name, address, state, city, phone, email, lat, long, bio } = vendor.attributes
    const formattedNum = formatPhoneNumber(phone)
    return {
      name,
      bio,
      address,
      state, 
      city,
      phone: formattedNum,
      email,
      lat,
      long, 
      id: parseInt(vendor.id)
    }
  })
}

export const cleanUser = (data, token) => {
  const products = cleanUserProducts(data.attributes.products)
  const { name, email, address, phone, bio, lat, long, city, state, zip, img_url } = data.attributes
  return {
    name, 
    email, 
    address, 
    phone: formatPhoneNumber(phone), 
    bio, 
    lat, 
    long, 
    city, 
    state, 
    zip, 
    img_url,
    products, 
    token
  }
}

export const cleanUserProducts = (products) => {
  const productArr = []
  products.reduce((obj, product) => {
    const itemName = Object.keys(product)[0];
    obj = {
      name: Object.keys(product)[0],
      ...product[itemName]
    }
    productArr.push(obj);
    return obj
  }, {})
  return productArr
}

export const cleanItems = (products) => {
  const productNames = products.reduce((arr, product) => {
    if(!arr.includes(product.name)) {
      arr.push(product.name)
    }
    return arr     
  },[]);

  const productIds = products.reduce((arr, product) => {
    if(!arr.includes(product.item_id)) {
      arr.push(product.item_id)
    }
    return arr     
  },[]);

  return productNames.map((name, i) => {
    return {
      name,
      id: productIds[i]
    }
  });
}