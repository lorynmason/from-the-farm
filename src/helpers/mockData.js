export const mockVendors = [
  {
    "id": "5",
    "type": "user",
    "attributes": {
        "name": "Basil Connection",
        "account_type": "vendor",
        "address": "4950 Beach Court",
        "city": "denver",
        "state": "CO",
        "zip": 80221,
        "email": "helter_skelter@example.com",
        "phone": 2313414141,
        "lat": 39.7861784,
        "long": -105.0178452,
        "bio": "The WORLDS Freshest Basil",
        "img_url": null,
        "products": [
            {
                "Berries": {
                    "id": 3,
                    "user_id": 5,
                    "item_id": 1,
                    "price": 1400,
                    "unit": "lb",
                    "description": "lb of berries"
                }
            },
            {
                "Potatoes": {
                    "id": 17,
                    "user_id": 5,
                    "item_id": 4,
                    "price": 170,
                    "unit": "lb",
                    "description": "a lb of potatoes"
                }
            }]
          }
  }
]

export const mockUser = {
  "id": "5",
  "type": "user",
  "attributes": {
    "name": "Basil Connection",
    "account_type": "vendor",
    "address": "4950 Beach Court",
    "city": "denver",
    "state": "CO",
    "zip": 80221,
    "email": "helter_skelter@example.com",
    "phone": 2313414141,
    "lat": 39.7861784,
    "long": -105.0178452,
    "bio": "The WORLDS Freshest Basil",
    "img_url": null,
    "products": [
      {
        "Berries": {
            "id": 3,
            "user_id": 5,
            "item_id": 1,
            "price": 1400,
            "unit": "lb",
            "description": "lb of berries"
        }
      },
      {
        "Potatoes": {
            "id": 17,
            "user_id": 5,
            "item_id": 4,
            "price": 170,
            "unit": "lb",
            "description": "a lb of potatoes"
        }
      }
    ]
  }
}
