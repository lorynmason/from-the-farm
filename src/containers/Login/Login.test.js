import React from 'react';
import { Login } from './Login';
import { shallow } from 'enzyme';
import { mockUser } from '../../helpers/mockData'

describe('Login', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Login user={mockUser} loginUser={jest.fn()}/>)
  })

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('handleChange', () => {
    it('should handle change', () => {
      expect(wrapper.state()).toEqual({
        email: '',
        password: '',
        name: '',
        password_confirmation: '',
        account_type: "vendor",  
        address: '',
        city: '',
        state: '',
        phone: '',
        zip: '',
        bio: '',
        img_url: '',
        newUser: false
      });

      wrapper.find('.login').simulate('change', {'target': {name: 'email', value: 'something@gmail.com'}});
      wrapper.find('.login').simulate('change', {'target': {name: 'password', value: 'grapes'}});
      

      expect(wrapper.state()).toEqual({
        email: 'something@gmail.com',
        password: 'grapes',
        name: '',
        password_confirmation: '',
        account_type: "vendor",  
        address: '',
        city: '',
        state: '',
        phone: '',
        zip: '',
        bio: '',
        img_url: '',
        newUser: false
      });
    })
  })
  describe('handleSubmit', () => {
    it('should call loginUser', () => {
      wrapper.find('.login').simulate('submit', {preventDefault: jest.fn()});
      expect(wrapper.instance().props.loginUser).toHaveBeenCalled()
    })
  })
})