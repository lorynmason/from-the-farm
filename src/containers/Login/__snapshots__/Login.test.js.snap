// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    loginUser={[MockFunction]}
    user={
      Object {
        "attributes": Object {
          "account_type": "vendor",
          "address": "4950 Beach Court",
          "bio": "The WORLDS Freshest Basil",
          "city": "denver",
          "email": "helter_skelter@example.com",
          "img_url": null,
          "lat": 39.7861784,
          "long": -105.0178452,
          "name": "Basil Connection",
          "phone": 2313414141,
          "products": Array [
            Object {
              "Berries": Object {
                "description": "lb of berries",
                "id": 3,
                "item_id": 1,
                "price": 1400,
                "unit": "lb",
                "user_id": 5,
              },
            },
            Object {
              "Potatoes": Object {
                "description": "a lb of potatoes",
                "id": 17,
                "item_id": 4,
                "price": 170,
                "unit": "lb",
                "user_id": 5,
              },
            },
          ],
          "state": "CO",
          "zip": 80221,
        },
        "id": "5",
        "type": "user",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3>
          Login
        </h3>,
        <input
          name="email"
          onChange={[Function]}
          placeholder="email"
          type="email"
          value=""
        />,
        <input
          name="password"
          onChange={[Function]}
          placeholder="password"
          type="password"
          value=""
        />,
        <button>
          Login
        </button>,
        undefined,
      ],
      "className": "login",
      "onChange": [Function],
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Login",
        },
        "ref": null,
        "rendered": "Login",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "name": "email",
          "onChange": [Function],
          "placeholder": "email",
          "type": "email",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "name": "password",
          "onChange": [Function],
          "placeholder": "password",
          "type": "password",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Login",
        },
        "ref": null,
        "rendered": "Login",
        "type": "button",
      },
      undefined,
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3>
            Login
          </h3>,
          <input
            name="email"
            onChange={[Function]}
            placeholder="email"
            type="email"
            value=""
          />,
          <input
            name="password"
            onChange={[Function]}
            placeholder="password"
            type="password"
            value=""
          />,
          <button>
            Login
          </button>,
          undefined,
        ],
        "className": "login",
        "onChange": [Function],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Login",
          },
          "ref": null,
          "rendered": "Login",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "email",
            "onChange": [Function],
            "placeholder": "email",
            "type": "email",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "password",
            "onChange": [Function],
            "placeholder": "password",
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Login",
          },
          "ref": null,
          "rendered": "Login",
          "type": "button",
        },
        undefined,
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
